incubator/binding-kafka-grpc/src/main/java/io/aklivity/zilla/runtime/binding/kafka/grpc/internal/stream/KafkaGrpcRemoteServerFactory.java

```diff
@@ -0,0 +1,1994 @@
+            final long budgetId = window.budgetId();
+            final int padding = window.padding();
+            final int capabilities = window.capabilities();
+
+            assert acknowledge <= sequence;
+            assert acknowledge >= delegate.replyAck;
+            assert maximum >= delegate.replyMax;
+
+            initialAck = acknowledge;
+            initialMax = maximum;
+            state = KafkaGrpcState.openInitial(state);
+
+            assert initialAck <= initialSeq;
+
+            delegate.onKafkaWindow(authorization, traceId, budgetId, padding, capabilities);
+        }
+
+        private void onKafkaReset(
+            ResetFW reset)
+        {
+            final long sequence = reset.sequence();
+            final long acknowledge = reset.acknowledge();
+            final long traceId = reset.traceId();
+            final long authorization = reset.authorization();
+
+            assert acknowledge <= sequence;
+            assert acknowledge >= delegate.initialAck;
+
+            delegate.initialAck = acknowledge;
+            state = KafkaGrpcState.closeInitial(state);
+
+            assert delegate.initialAck <= delegate.initialSeq;
+
+            delegate.onKafkaReset(traceId, authorization);
+
+            doKafkaReset(traceId, authorization);
+        }
+
+        private void doKafkaReset(
+            long traceId,
+            long authorization)
+        {
+            if (KafkaGrpcState.replyOpened(state) &&
+                !KafkaGrpcState.replyClosed(state))
+            {
+                state = KafkaGrpcState.closeReply(state);
+
+                doReset(kafka, originId, routedId, replyId, replySeq, replyAck, replyMax,
+                    traceId, authorization, emptyRO);
+            }
+        }
+
+        private void doKafkaWindow(
+            long traceId)
+        {
+            if (kafka != null && !KafkaGrpcState.replyClosed(state))
+            {
+                doWindow(kafka, originId, routedId, replyId, replySeq, replyAck, replyMax,
+                    traceId, 0L, replyBud, replyPad, replyCap);
+            }
+        }
+
+        private void doKafkaTombstone(
+            long traceId,
+            long authorization,
+            String16FW status)
+        {
+            Flyweight tombstoneDataEx = kafkaDataExRW
+                .wrap(extBuffer, 0, extBuffer.capacity())
+                .typeId(kafkaTypeId)
+                .merged(m -> m
+                    .timestamp(now().toEpochMilli())
+                    .partition(p -> p.partitionId(-1).partitionOffset(-1))
+                    .key(k -> result.key(delegate.correlationId, k))
+                    .headers(h -> result.headersWithStatusCode(delegate.correlationId, status, h)))
+                .build();
+
+            doKafkaData(traceId, authorization, delegate.initialBud, 0, DATA_FLAG_COMPLETE, null, tombstoneDataEx);
+        }
+    }
+
+    private final class KafkaErrorProducer
```

Comment:
Does this need to be a full class?

Seems like we only use `KafkaErrorProducer.doKafkaTombstone(...)` method.

Reply:
Yeah but we are going to use KafkaCorrelateProxy it should have grpc client. We need this because grpc client may not have initialize yet.

Reply:
I see, the call to `doKafkaDataNull` is detecting that `doKafkaBegin` is needed first, then registering the `onKafkaMessage` callback to open the stream before sending the tombstone.

Does this work if you have multiple different Kafka messages with different correlation ids that need to be rejected?

I think the concept of `lastCorrelationId` will not work in `KafkaErrorProducer` as it will be overwritten if there are multiple `doKafkaDataNull` calls before `onKafkaWindow`, agree?

Reply:
Yep!